[google-dawn] new port

Arcnor opened the pull request:
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [ ] The versioning scheme in `vcpkg.json` matches what upstream says.
- [ ] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

Regarding "versioning scheme", there is no specific versioning for this library, so a short SHA1 hash has been used instead.
Regarding "license declaration", I've added the LICENSE as copyright, not sure if there anything else to do about it.

Fixes #22948

Arcnor commented:
@microsoft-github-policy-service agree

Arcnor commented:
The Android builds seem to be failing with an error about not being able to find `jinja2`, but I'm using `vcpkg-get-python-packages` like with the other platforms, not sure what else needs to happen there...

Adela0814 commented on ports/google-dawn/vcpkg.json:
```diff
@@ -0,0 +1,59 @@
+{
+  "name": "google-dawn",
+  "version-string": "e3feea76729d",
```
Maybe you can change the version to chromium_5851?

Adela0814 commented on ports/google-dawn/absl_lts.diff:
```diff
@@ -0,0 +1,569 @@
+diff --git a/generator/templates/dawn/native/api_absl_format.cpp b/generator/templates/dawn/native/api_absl_format.cpp
+index a3b7ea22c..19fa467da 100644
+--- a/generator/templates/dawn/native/api_absl_format.cpp
++++ b/generator/templates/dawn/native/api_absl_format.cpp
```
This patch content should be submitted to the upstream first. We need to wait for the upstream to release a new version or agree to this change.

Arcnor replied:
I might be mistaken, but this seems to be an artifact of using vcpkg's `absl` port instead of using HEAD directly, like upstream does, so I don't think they will want to change this.

How do we make them "agree" to this change?

Adela0814 commented on ports/google-dawn/portfile.cmake:
```diff
@@ -0,0 +1,51 @@
+vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
+
+vcpkg_from_git(
+		OUT_SOURCE_PATH SOURCE_PATH
+		URL https://dawn.googlesource.com/dawn
+		REF e3feea76729d52c050908e5f8588ee0b964112ff
+		PATCHES cmake_changes.diff absl_lts.diff src_tint_prefix.diff
```
Please modify all formats in this file. 
```suggestion
vcpkg_from_git(
	OUT_SOURCE_PATH SOURCE_PATH
	URL https://dawn.googlesource.com/dawn
	REF e3feea76729d52c050908e5f8588ee0b964112ff
	PATCHES cmake_changes.diff absl_lts.diff src_tint_prefix.diff
```

Adela0814 commented:
Note: I will be converting your PR to draft status. When you respond, please revert to "ready for review". That way, I can be aware that you've responded since you can't modify the tags.

sshsphere commented:
Any updates on this?

Arcnor commented:
@Adela0814 sorry for the delay, but I've updated Dawn again and made a different change for the Abseil issue, hopefully this one is easier (it's just removing a forward declaration this time, although I'm still not sure why this happens).

I've also set the version to `chromium_6097` like you suggested.

Adela0814 commented on ports/google-dawn/absl_lts.diff:
```diff
@@ -0,0 +1,15 @@
+@@ -36,10 +36,6 @@
+ #include "dawn/common/RefCounted.h"
+ #include "dawn/native/Forward.h"
+
+-namespace absl {
+-class FormatSink;
+-}
+-
```
Can you explain why this was deleted? This modification should be submitted to the upstream to let them know.

Arcnor replied:
This is only required because when adding `vcpkg`'s absl, there is a conflict for some reason. Removing this single forward declaration takes care of the problem. This won't be accepted upstream, as it's only an issue with `vcpkg`.

Adela0814 commented on ports/google-dawn/portfile.cmake:
```diff
@@ -0,0 +1,51 @@
+
+vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
+
+vcpkg_from_git(
+	OUT_SOURCE_PATH SOURCE_PATH
+	URL https://dawn.googlesource.com/dawn
+	REF 52564ab1a0e3359b7bbadd282865cc28612c33ce
+	PATCHES cmake_changes.diff absl_lts.diff src_tint_prefix.diff
+)
```
```suggestion
vcpkg_from_git(
    OUT_SOURCE_PATH SOURCE_PATH
    URL https://dawn.googlesource.com/dawn
    REF 52564ab1a0e3359b7bbadd282865cc28612c33ce
    PATCHES 
        cmake_changes.diff
        absl_lts.diff
        src_tint_prefix.diff
)
```

Adela0814 commented on ports/google-dawn/portfile.cmake:
```diff
@@ -0,0 +1,51 @@
+	PATCHES cmake_changes.diff absl_lts.diff src_tint_prefix.diff
+)
+
+vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+	FEATURES
+	opengl DAWN_ENABLE_DESKTOP_GL
+	opengles DAWN_ENABLE_OPENGLES
+	glfw DAWN_USE_GLFW
+)
```
```suggestion
vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
    FEATURES
    opengl DAWN_ENABLE_DESKTOP_GL
    opengles DAWN_ENABLE_OPENGLES
    glfw DAWN_USE_GLFW
)
```

Adela0814 commented on ports/google-dawn/portfile.cmake:
```diff
@@ -0,0 +1,51 @@
+
+file(COPY "${CMAKE_CURRENT_LIST_DIR}/Config.cmake.in" DESTINATION "${SOURCE_PATH}")
+
+vcpkg_cmake_configure(
+	SOURCE_PATH "${SOURCE_PATH}"
+	OPTIONS
+	    -DDAWN_ENABLE_VULKAN=OFF
+	    -DDAWN_BUILD_SAMPLES=OFF
+	    -DTINT_BUILD_TESTS=OFF
+	    -DTINT_BUILD_CMD_TOOLS=OFF
+	    -DTINT_BUILD_DOCS=OFF
+	    ${FEATURE_OPTIONS}
+)
```
```suggestion
vcpkg_cmake_configure(
    SOURCE_PATH "${SOURCE_PATH}"
    OPTIONS
        -DDAWN_ENABLE_VULKAN=OFF
        -DDAWN_BUILD_SAMPLES=OFF
        -DTINT_BUILD_TESTS=OFF
        -DTINT_BUILD_CMD_TOOLS=OFF
        -DTINT_BUILD_DOCS=OFF
        ${FEATURE_OPTIONS}
)
```

Adela0814 commented on ports/google-dawn/portfile.cmake:
```diff
@@ -0,0 +1,51 @@
+)
+
+vcpkg_cmake_install()
+vcpkg_fixup_cmake_targets(
+	CONFIG_PATH "cmake"
+)
```
```suggestion
vcpkg_fixup_cmake_targets(
    CONFIG_PATH "cmake"
)
```

Adela0814 commented on ports/google-dawn/portfile.cmake:
```diff
@@ -0,0 +1,51 @@
+
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+
+file(
+	INSTALL "${SOURCE_PATH}/LICENSE"
+	DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}"
+	RENAME copyright
+)
```
```suggestion
vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE")
```

Adela0814 commented on ports/google-dawn/portfile.cmake:
```diff
@@ -0,0 +1,51 @@
+	RENAME copyright
+)
+
+file(
+	INSTALL "${CMAKE_CURRENT_LIST_DIR}/usage"
+	DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}"
+)
```
```suggestion
file(
    INSTALL "${CMAKE_CURRENT_LIST_DIR}/usage"
    DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}"
)
```

Adela0814 commented on ports/google-dawn/src_tint_prefix.diff:
Your patch is very large. Is it possible to deploy these modifications upstream and release them?

Arcnor replied:
I've asked them, and they don't want to do the changes, so it's not possible.

Arcnor replied:
That said, this patch, albeit large, it's only doing a single thing: replacing all `include 'src/tint/...'` by `include 'tint/...'` so it works properly after CMake install.

dg0yt commented on ports/google-dawn/portfile.cmake:
```diff
@@ -0,0 +1,47 @@
+vcpkg_add_to_path("${PYTHON3_DIR}")
+x_vcpkg_get_python_packages(PYTHON_EXECUTABLE "${PYTHON3}" PACKAGES jinja2)
+
+vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
```
Is this a windows-only restriction?

Arcnor replied:
I'm not sure about this, it's been a long time since I compiled on another platform, but Google said a while ago "Ideally the vast majority of dawn_native should be compiled as a static lib, that's linked in either dawn_native.dll or webgpu_dawn.dll (when building with shared libs)", but that to do that it will require changing how the compilation works. I'm not ready to do that kind of changes, or even sure if that will only affect Windows only or other platforms too, so I think we should get the first package out there if possible, then iterate over it, it's been painful enough as it is :).

dg0yt commented on ports/google-dawn/portfile.cmake:
```diff
@@ -0,0 +1,47 @@
+)
+
+vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+    FEATURES
+    opengl DAWN_ENABLE_DESKTOP_GL
+    opengles DAWN_ENABLE_OPENGLES
+    glfw DAWN_USE_GLFW
```
```suggestion
    FEATURES
        glfw     DAWN_USE_GLFW
        opengl   DAWN_ENABLE_DESKTOP_GL
        opengles DAWN_ENABLE_OPENGLES
```

dg0yt commented on ports/google-dawn/portfile.cmake:
```diff
@@ -0,0 +1,47 @@
+)
+
+vcpkg_cmake_install()
+vcpkg_fixup_cmake_targets(
+    CONFIG_PATH "cmake"
+)
```
```suggestion
vcpkg_fixup_cmake_targets(CONFIG_PATH "cmake")
```

dg0yt commented on ports/google-dawn/usage:
```diff
@@ -0,0 +1,7 @@
+The package dawn provides CMake targets:
```
```suggestion
dawn provides CMake targets:
```

Arcnor replied:
`vcpkg` itself generates "The package XXX provides CMake targets:`, do you want me to deviate from that?

dg0yt reviewed (commented):
IMO the CMake patches are too invasive.
I don't think we can provide an exported version (`1.0.0`) when upstream doesn't provide this version.
I don't understand why the patch needs to introduce `ALIAS` targets.
I didn't look any further.

Adela0814 commented on ports/google-dawn/vcpkg.json:
```diff
@@ -0,0 +1,59 @@
+{
+  "name": "dawn",
```
```suggestion
  "name": "google-dawn",
```

Arcnor commented:
> IMO the CMake patches are too invasive.
Can you elaborate a bit more? Are you talking about the big patch (`src_tint_prefix.diff`) to fix the imports? That one can be removed, although `tint` will probably be unusable then, but Dawn will continue to work, unless somebody has a better idea. Google says they want to fix this at some indeterminate point in the future by actually moving the files themselves, but that I cannot do myself.

> I don't think we can provide an exported version (`1.0.0`) when upstream doesn't provide this version.
If you mean the CMake project version, I just added that to keep CMake happy, I thought it was inconsequential given that vcpkg has its own versioning. What do you suggest to have there instead? What you folks suggested was having `chromium_6097` which cannot be a valid CMake project version, so I probably need something different there.

> I don't understand why the patch needs to introduce `ALIAS` targets.
It doesn't, I just added them for convenience, didn't think those would be a problem.

> I didn't look any further.
Appreciate you taking the time, thanks!

Arcnor commented:
@Adela0814 I'm going to add the other suggestions you've folks have made, but I'm not sure if you're happy with me removing and force pushing those two commits all the time? I'm assuming I cannot just simply "Apply suggestion" here in GH as I still need to fix the git-tree version after doing any changes...

Adela0814 commented:
> @Adela0814 I'm going to add the other suggestions you've folks have made, but I'm not sure if you're happy with me removing and force pushing those two commits all the time? I'm assuming I cannot just simply "Apply suggestion" here in GH as I still need to fix the git-tree version after doing any changes...

If removing your big patch won't affect the down usage, I think it's best not to upload such a large modification. We can update it later when there are relevant upstream fixes.

Arcnor committed:
[google-dawn] new port

Arcnor committed:
[google-dawn] update version

Arcnor commented:
> > @Adela0814 I'm going to add the other suggestions you've folks have made, but I'm not sure if you're happy with me removing and force pushing those two commits all the time? I'm assuming I cannot just simply "Apply suggestion" here in GH as I still need to fix the git-tree version after doing any changes...
> 
> If removing your big patch won't affect the `dawn` usage, I think it's best not to upload such a large modification. We can update it later when there are relevant upstream fixes.

Ok, I've removed tint, let me know if there is anything else.

Adela0814 commented:
Please fix the CI error, find details from https://dev.azure.com/vcpkg/public/_build/results?buildId=96439&view=artifacts&pathAsName=false&type=publishedArtifacts.

Error message:
```
Line   7: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\constant\manager.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/constant/manager.h': No such file or directory
	Line  11: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\constant\node.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/constant/node.h': No such file or directory
	Line  15: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\api\tint.cc(28): fatal error C1083: Cannot open include file: 'src/tint/api/tint.h': No such file or directory
	Line  19: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\constant\scalar.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/constant/scalar.h': No such file or directory
	Line  23: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\constant\composite.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/constant/composite.h': No such file or directory
	Line  27: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\constant\eval.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/constant/eval.h': No such file or directory
	Line  31: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\intrinsic\ctor_conv.cc(37): fatal error C1083: Cannot open include file: 'src/tint/lang/core/intrinsic/ctor_conv.h': No such file or directory
	Line  36: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\constant\value.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/constant/value.h': No such file or directory
	Line  40: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\clone_context.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/clone_context.h': No such file or directory
	Line  44: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\block.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/block.h': No such file or directory
	Line  48: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\intrinsic\table.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/intrinsic/table.h': No such file or directory
	Line  52: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\constant\splat.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/constant/splat.h': No such file or directory
	Line  56: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\binary.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/binary.h': No such file or directory
	Line  60: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\constant.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/constant.h': No such file or directory
	Line  64: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\block_param.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/block_param.h': No such file or directory
	Line  68: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\access.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/access.h': No such file or directory
	Line  72: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\bitcast.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/bitcast.h': No such file or directory
	Line  76: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\continue.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/continue.h': No such file or directory
	Line  80: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\control_instruction.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/control_instruction.h': No such file or directory
	Line  84: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\call.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/call.h': No such file or directory
	Line  88: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\builder.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/builder.h': No such file or directory
	Line  92: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\break_if.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/break_if.h': No such file or directory
	Line  96: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\construct.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/construct.h': No such file or directory
	Line 100: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\core_builtin_call.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/core_builtin_call.h': No such file or directory
	Line 104: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\disassembler.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/disassembler.h': No such file or directory
	Line 108: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\discard.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/discard.h': No such file or directory
	Line 112: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\exit.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/exit.h': No such file or directory
	Line 116: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\convert.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/convert.h': No such file or directory
	Line 120: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\exit_if.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/exit_if.h': No such file or directory
	Line 124: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\ir\builtin_call.cc(28): fatal error C1083: Cannot open include file: 'src/tint/lang/core/ir/builtin_call.h': No such file or directory
	Line 129: D:\buildtrees\google-dawn\src\28612c33ce-d03c9c5de7.clean\src\tint\lang\core\intrinsic\data.cc(40): fatal error C1083: Cannot open include file: 'src/tint/lang/core/intrinsic/dialect.h': No such file or directory
```

Adela0814 commented:
Closing this PR since it seems that no progress is being made. Please ping us to reopen if work is still being done.

retifrav commented:
> the CMake patches are too invasive

Sadly, that's just how things are with this particular project. If only Dawn authors/maintainers knew how to install targets and make proper CMake packages.

> I don't understand why the patch needs to introduce ALIAS targets

As I understand it, that's for having nice "namespaced" targets for linking, for example `target_link_libraries(main PRIVATE dawn::cpp dawn::native)`. Many projects have it this way in their CMake packages.

Either way, big thanks to @Arcnor for starting this work. Without his patches I would spend much more time figuring out the targets relations in this project. I am still not sure if I did figure them out completely, but at least I got to the point where it works.

I've created [my own port for Dawn](https://github.com/retifrav/vcpkg-registry/tree/master/ports/dawn) in my personal vcpkg registry. There are likely things that I've missed and stuff to improve still, but it seems to build fine (*tested on Windows, Mac OS and GNU/Linux so far*), and consuming projects are able to use it.

I doubt that my port will soon (*if ever*) be ready to be submitted to the Microsoft's registry (*at the very least because it depends on my custom versions of other ports*), but I thought I'll mention it here anyway, in case others would like to re-create it in their own registries.

JackBoosY commented:
I request to use this PR changes to add this port.

dg0yt commented:
FTR we do build dawn as part of skia, using gn. We also export unofficial CMake config from the skia gn configuration.
So IMO the natural step for the official registry would be to factor out dawn from skia.

Arcnor commented:
> FTR we do build dawn as part of skia, using gn. We also export unofficial CMake config from the skia gn configuration. So IMO the natural step for the official registry would be to factor out dawn from skia.

You do build it as part of Skia, but IIRC there will be no API exported if you do that (many of the changes here that you folks didn't like are just to get the API useable by exposing the includes, etc). Also, using an unofficial CMake file when there is an official one upstream seems wrong somehow, but if it gets it faster on the registry that'd be cool.

dg0yt commented:
> many of the changes here that you folks didn't like are just to get the API useable by exposing the includes, etc

The changes that I didn't like here where those that weren't necessary, in particular the ALIAS targets.

And you got reviews because people wanted to help form into a mergable and maintainable state. It is not a trivial port.

Arcnor commented:
> > many of the changes here that you folks didn't like are just to get the API useable by exposing the includes, etc
> 
> The changes that I didn't like here where those that weren't necessary, in particular the ALIAS targets.

I'm not talking only about your comments, many of the iterations here were related to the intrusive includes, and that's what I'm referring to.

> And you got reviews because people wanted to help form the PR into a mergable and maintainable state. It is not a trivial port.

I understand that, and I still was thankful for you for taking the time to review it, I'm not sure that came across above, but I was :).

But not sure what getting reviews has to do with my comment above, I think it's still true, both the part about the includes (unless upstream has changed that, or the Skia changes already include something for it) as well as using a different CMake.

In any case, I hope you folks get it into a finished state no matter how, I'm sure people will be thankful for that.
